File: marble-backoffice/src/models/RuleDto.ts
import * as yup from "yup";
import { adaptDtoWithYup } from "@/infra/adaptDtoWithYup";
import type { Rule } from "./Rule";
import { adaptAstNode, AstNodeSchemaNullable } from "./AstExpressionDto";

export const RuleSchema = yup.object({
  id: yup.string().required(),
  scenarioIterationId: yup.string().required(),
  displayOrder: yup.number().defined(),
  name: yup.string().defined(),
  description: yup.string().defined(),
  formula_ast_expression: AstNodeSchemaNullable,
  scoreModifier: yup.number().defined(),
  createdAt: yup.date().required(),
});

export type RuleSchemaDto = yup.InferType<typeof RuleSchema>;

export function adaptRule(dto: RuleSchemaDto): Rule {
  return {
    ruleId: dto.id,
    iterationId: dto.scenarioIterationId,
    displayOrder: dto.displayOrder,
    name: dto.name,
    description: dto.description,
    formulaAstExpression:
      dto.formula_ast_expression === null
        ? null
        : adaptAstNode(dto.formula_ast_expression),
    scoreModifier: dto.scoreModifier,
    createdAt: dto.createdAt,
  };
}

export function adaptRuleApiResult(json: unknown): Rule {
  const dto = adaptDtoWithYup(json, yup.object({ rule: RuleSchema }));
  return adaptRule(dto.rule);
}

export function adaptUpdateRuleApiResult(json: unknown): Rule {
  const dto = adaptDtoWithYup(
    json,
    yup.object({
      rule: RuleSchema,
    })
  );
  return adaptRule(dto.rule);
}

File: dto/rule_dto.go
package dto

import (
	"fmt"
	"time"

	"github.com/checkmarble/marble-backend/models"
)

type RuleDto struct {
	Id                   string    `json:"id"`
	ScenarioIterationId  string    `json:"scenario_iteration_id"`
	DisplayOrder         int       `json:"display_order"`
	Name                 string    `json:"name"`
	Description          string    `json:"description"`
	FormulaAstExpression *NodeDto  `json:"formula_ast_expression"`
	ScoreModifier        int       `json:"score_modifier"`
	CreatedAt            time.Time `json:"created_at"`
	RuleGroup            string    `json:"rule_group"`
}

type RuleExecutionData struct {
	Version string `json:"version"`
	Status  string `json:"status"`
	Total   int    `json:"total"`
}

type CreateRuleInputBody struct {
	ScenarioIterationId  string   `json:"scenario_iteration_id"`
	DisplayOrder         int      `json:"display_order"`
	Name                 string   `json:"name"`
	Description          string   `json:"description"`
	FormulaAstExpression *NodeDto `json:"formula_ast_expression"`
	ScoreModifier        int      `json:"score_modifier"`
	RuleGroup            string   `json:"rule_group"`
}

type UpdateRuleBody struct {
	DisplayOrder         *int     `json:"display_order,omitempty"`
	Name                 *string  `json:"name,omitempty"`
	Description          *string  `json:"description,omitempty"`
	FormulaAstExpression *NodeDto `json:"formula_ast_expression"`
	ScoreModifier        *int     `json:"score_modifier,omitempty"`
	RuleGroup            *string  `json:"rule_group"`
}

func AdaptRuleDto(rule models.Rule) (RuleDto, error) {
	var formulaAstExpression *NodeDto
	if rule.FormulaAstExpression != nil {
		nodeDto, err := AdaptNodeDto(*rule.FormulaAstExpression)
		if err != nil {
			return RuleDto{}, err
		}
		formulaAstExpression = &nodeDto
	}

	return RuleDto{
		Id:                   rule.Id,
		ScenarioIterationId:  rule.ScenarioIterationId,
		DisplayOrder:         rule.DisplayOrder,
		Name:                 rule.Name,
		Description:          rule.Description,
		FormulaAstExpression: formulaAstExpression,
		ScoreModifier:        rule.ScoreModifier,
		CreatedAt:            rule.CreatedAt,
		RuleGroup:            rule.RuleGroup,
	}, nil
}

func ProcessRuleExecutionDataDtoFromModels(inputs []models.RuleExecutionStat) []RuleExecutionData {
	result := make([]RuleExecutionData, len(inputs))
	for i, input := range inputs {
		item := RuleExecutionData{
			Version: input.Version,
			Status:  input.Outcome,
			Total:   input.Total,
		}
		result[i] = item
	}
	return result
}

func AdaptCreateRuleInput(body CreateRuleInputBody, organizationId string) (models.CreateRuleInput, error) {
	createRuleInput := models.CreateRuleInput{
		OrganizationId:       organizationId,
		ScenarioIterationId:  body.ScenarioIterationId,
		DisplayOrder:         body.DisplayOrder,
		Name:                 body.Name,
		Description:          body.Description,
		FormulaAstExpression: nil,
		ScoreModifier:        body.ScoreModifier,
		RuleGroup:            body.RuleGroup,
	}

	if body.FormulaAstExpression != nil {
		node, err := AdaptASTNode(*body.FormulaAstExpression)
		if err != nil {
			return models.CreateRuleInput{}, fmt.Errorf(
				"could not adapt formula ast expression: %w %w", err, models.BadParameterError)
		}
		createRuleInput.FormulaAstExpression = &node
	}

	return createRuleInput, nil
}

func AdaptUpdateRule(ruleId string, body UpdateRuleBody) (models.UpdateRuleInput, error) {
	updateRuleInput := models.UpdateRuleInput{
		Id:                   ruleId,
		DisplayOrder:         body.DisplayOrder,
		Name:                 body.Name,
		Description:          body.Description,
		FormulaAstExpression: nil,
		ScoreModifier:        body.ScoreModifier,
		RuleGroup:            body.RuleGroup,
	}

	if body.FormulaAstExpression != nil {
		node, err := AdaptASTNode(*body.FormulaAstExpression)
		if err != nil {
			return models.UpdateRuleInput{}, fmt.Errorf(
				"could not adapt formula ast expression: %w %w", err, models.BadParameterError)
		}
		updateRuleInput.FormulaAstExpression = &node
	}

	return updateRuleInput, nil
}
